src/Orleans/Streams/PubSub/ImplicitStreamSubscriberTable.cs

```diff
@@ -51,31 +58,39 @@ internal void InitImplicitStreamSubscribers(IEnumerable<Type> grainClasses)
         /// <exception cref="System.InvalidOperationException">Internal invariant violation.</exception>
         internal IDictionary<Guid, IStreamConsumerExtension> GetImplicitSubscribers(StreamId streamId, IInternalGrainFactory grainFactory)
         {
-            if (String.IsNullOrWhiteSpace(streamId.Namespace))
+            if (string.IsNullOrWhiteSpace(streamId.Namespace))
             {
                 throw new ArgumentException("The stream ID doesn't have an associated namespace.", nameof(streamId));
             }
 
-            HashSet<int> entry;
+            HashSet<int> entry = GetOrAddImplicitSubscribersSet(streamId.Namespace);
+
             var result = new Dictionary<Guid, IStreamConsumerExtension>();
-            if (table.TryGetValue(streamId.Namespace, out entry))
+            foreach (var i in entry)
             {
-                foreach (var i in entry)
+                IStreamConsumerExtension consumer = MakeConsumerReference(grainFactory, streamId.Guid, i);
+                Guid subscriptionGuid = MakeSubscriptionGuid(i, streamId);
+                if (result.ContainsKey(subscriptionGuid))
                 {
-                    IStreamConsumerExtension consumer = MakeConsumerReference(grainFactory, streamId.Guid, i);
-                    Guid subscriptionGuid = MakeSubscriptionGuid(i, streamId);
-                    if (result.ContainsKey(subscriptionGuid))
-                    {
-                        throw new InvalidOperationException(string.Format("Internal invariant violation: generated duplicate subscriber reference: {0}, subscriptionId: {1}", consumer, subscriptionGuid));
-                    }
-                    result.Add(subscriptionGuid, consumer);
-                }                
-                return result;                
+                    throw new InvalidOperationException(
+                        $"Internal invariant violation: generated duplicate subscriber reference: {consumer}, subscriptionId: {subscriptionGuid}");
+                }
+                result.Add(subscriptionGuid, consumer);
             }
-
             return result;
         }
 
+        private HashSet<int> GetOrAddImplicitSubscribersSet(string streamNamespace)
+        {
+            HashSet<int> entry;
+            if (!table.TryGetValue(streamNamespace, out entry))
+            {
+                entry = FindImplicitSubscribers(streamNamespace);
+                table[streamNamespace] = entry;
```

Comment:
Since table is built at runtime, rather than at start, it needs to be a concurrent dictionary, correct?  Or is their some other concurrency guarantee that I'm missing?